string code
---
bool success
---
char key
string code_typed
string state
